app/_locales/en/messages.json

```
@@ -637,6 +637,10 @@
   "gasPriceNoDenom": {
     "message": "Gas Price"
   },
+  "gdprMessage": {
+    "message": "This data is aggregated and is therefore anonymous for the purposes of General Data Protection Regulation (EU) 2016/679. For more information in relation to our privacy practices, please see our {{Privacy Policy here}}.",
```

Using double-curly-braces as the marker for an embedded React component makes sense, but why the name? Using a number instead would allow embedding the same component in multiple places in the message, which is something we've needed to do with the normal substitutions before (e.g. if we want a word to be a link, and it appears twice)

---

Oh I see - the string content is wrapped by the React component itself, and is part of the translated message.

This seems a bit more complex than it needs to be. Not every embedded React component will need to wrap text, and those that do can use a separate localized message instead of an embedded one. It would also make the implementation simpler, as you could just... use the real React element passed in rather than cloning it with `cloneElement`

---

Yeah, could this lean on the existing substitutions, allowing for a substitution to be a React component:

```suggestion
    "message": "This data is aggregated and is therefore anonymous for the purposes of General Data Protection Regulation (EU) 2016/679. For more information in relation to our privacy practices, please see our $1.",
```

And we'd use it like so:

```js
{ t('gdprMessage', [(
  <a
    key="metametrics-bottom-text-wrapper"
    href="https://metamask.io/privacy.html"
    target="_blank"
    rel="noopener noreferrer"
  >
  { t('privacyPolicy') }
  </a>
)])}
```

---

My concern here is that not all translations are structured in the same way, and many languages have very different semantics from each other. Consider for example these different translations of the `"blockExplorerView" message:

```
_locales/en/messages.json:
"message": "View account at $1",
"description": "$1 replaced by URL for custom block explorer"

_locales/am/messages.json:
  "message": "መለያን በ$1እይ",

_locales/ar/messages.json:
"message": "عرض الحساب في $1",

_locales/bg/messages.json:
"message": "Преглед на акаунт на $1",

_locales/bn/messages.json:
"message": "$1 এ অ্যাকাউন্ট দেখুন",

_locales/ca/messages.json:
"message": "Veure compte a $1 "

_locales/de/messages.json:
"Konto bei $1 anzeigen",
```

The substitution can be placed at various points in the string. And the order, grammar and meaning of the surrounding words likely varies.

A phrase like "Privacy Policy here" may not stay as a single unit/sequence in different languagaes.

And, perhaps more significantly, the translation of that phrase could differ in different contexts. Exactly what "Privacy policy here" means as a stand alone phrase may not even be entirely clear to an english speaker. In other languages, how that phrase is translated could vary based on the context it is placed in. So we can simply compose smaller translations into larger translations and hope they are coherent. We have to leave that work to the translator.

For the case of substitutions that will remain constant across any language - an account address or name for example - we don't have a problem and can just wrap those items in a component similar to Whymarrh's example above.

But I think that for cases where words or phrases need to be translated and be wrapped in a component for specific styling or behaviour, then we can't simply compose two translations into one. We have to ensure the translator knows the context in which the smaller phrase (the one to be wrapped in the component) needs to be used.

---

>We have to ensure the translator knows the context in which the smaller phrase (the one to be wrapped in the component) needs to be used.

But we can do that, right? That's what the `description` field is for - we can tell the translator exactly how and where it's used.

I agree with all of what you've said here but I don't understand how it's applicable to what Whymarrh and I are suggesting.

---

So to make sure I am clear on what you are saying, you are suggesting that (to take Whymarrh's example) we would:
- reference 'gdprMessage' in the description of the 'privacyPolicy', saying that this translation is intended to exclusively be used as a replacement for $1 in 'gdprMessage', and then the translator would look that message up
- reference 'privacyPolicy' in the description of 'gdprMessage', saying that $1 will be replaced by the translation of the 'privaclyPolicy' message
- and then also include both 'gdprMessage' and 'privacyPolicy' in the code of the react component that needs 'gdprMessage'

---

Yep, that sounds about right to me!

Though I would keep the messages together; e.g. name it `gdprMessagePrivacyPolicy` instead, so it's right next to it. I've tried to do that for any related groups of messages.

---

That way we could accommodate cases where:

* The same React substitution is used multiple times
* A React substitution doesn't wrap any text at all
* A React substitution wraps multiple distinct pieces of text

---

okay, I agree that is better, I will change this to take that approach

---

I have made the changes according to the above discussion in db5243d7b